framework/utils/randomUtils.py

```diff
@@ -42,17 +43,22 @@ def __init__(self):
       @ Out, None
     """
     self.queue = defaultdict(deque)
+    self.__queueLock = threading.RLock()
 
   def generate(self,engine=None):
     """
       Yields a normally-distributed pseudorandom value
       @ In, engine, instance, optional, random number generator
       @ Out, generate, float, random value
     """
-    if len(self.queue[engine]) == 0:
-      #calculate new values
-      self.queue[engine].extend(self.createSamples(engine=engine))
-    return self.queue[engine].pop() #no need to pop left, as they're independent and all get used
+    with self.__queueLock:
```

Comment:
is the lock required? if it is because of the usage in a "parallelized" model, shouldn't it be handled outside?

Reply:
We don't know when an outside set of entities will attempt to access the RNG engine either in serial or parallel, and it can be accessed by many different entities. We were seeing a lockup where two entities in parallel were trying to access the same sample from Box Muller, and resulted in an empty queue error. Since the BoxMuller is aware of its limitation, but the outside entity isn't, this seems like a more stable and convenient solution to me so far.